### PR TITLE
feat: integrate stableswap pallet

### PR DESCRIPTION
**Pull request by enthusiastmartin**

### CODE REVIEW

File: runtime/hydradx/src/assets.rs [hunk 1]
```diff
@@ -480,3 +484,47 @@ impl pallet_dynamic_fees::Config for Runtime {
...
+{
+	type AccountId = T::AccountId;
+
+	fn from_assets(asset: &AssetId, identifier: Option<&[u8]>) -> Self::AccountId {
+		let name = Self::name(asset, identifier);
+		T::AccountId::unchecked_from(<T::Hashing as frame_support::sp_runtime::traits::Hash>::hash(&name[..]))
+	}
+
+	fn name(asset: &u32, identifier: Option<&[u8]>) -> Vec<u8> {
+		let mut buf: Vec<u8> = if let Some(ident) = identifier {
```
Comment (Roznovjak) [inline runtime/hydradx/src/assets.rs:506]: a nitpick: `let mut buf: Vec<u8> = identifier.map_or_else(|| vec![], |v| v.to_vec());`

File: pallets/stableswap/src/weights.rs [hunk 2]
```diff
@@ -55,81 +60,91 @@ pub trait WeightInfo {
 	fn update_amplification() -> Weight;
 }
 
-pub struct BasiliskWeight<T>(PhantomData<T>);
+pub struct SubstrateWeight<T>(PhantomData<T>);
```
Comment (Roznovjak) [inline pallets/stableswap/src/weights.rs:63]: How are these weights different from `HydraWeight`?

Comment (enthusiastmartin) [reply pallets/stableswap/src/weights.rs:63]: No different. HydraWeights name is used in the runtime. Fort the default, i'd like to have it named SubstrateWeight.

File: pallets/dca/Cargo.toml [hunk 3]
```diff
@@ -34,7 +34,9 @@ hydradx-traits = { workspace = true }
 hydradx-adapters = { workspace = true }
 pallet-relaychain-info = { workspace = true }
 pallet-ema-oracle =  { workspace = true }
+pallet-stableswap =  { workspace = true }
```
Comment (mrq1911) [inline pallets/dca/Cargo.toml:37]: DCA shoudn't be dependent on stableswap at all

File: runtime/hydradx/src/assets.rs [hunk 4]
```diff
@@ -486,6 +488,47 @@ impl pallet_dynamic_fees::Config for Runtime {
...
+	type AssetId = AssetId;
+	type Currency = Currencies;
+	type ShareAccountId = StableswapAccountIdConstructor<Runtime>;
+	type AssetInspection = AssetRegistry;
+	type AuthorityOrigin = EnsureRoot<AccountId>;
+	type DustAccountHandler = Duster;
+	type MinPoolLiquidity = MinPoolLiquidity;
+	type MinTradingLimit = MinTradingLimit;
+	#[cfg(feature = "runtime-benchmarks")]
+	type BenchmarkHelper = (); //TODO: this must be some actual implementation
```
Comment (mrq1911) [inline runtime/hydradx/src/assets.rs:527]: missing implementation

Comment (mrq1911) [inline runtime/hydradx/src/assets.rs:527]: missing impl

File: traits/Cargo.toml [hunk 5]
```diff
@@ -1,6 +1,6 @@
 [package]
 name = "hydradx-traits"
-version = "2.6.0"
+version = "2.6.1"
```
Comment (mrq1911) [inline traits/Cargo.toml:3]: bump version not necessary

**Review by mrq1911 — CHANGES_REQUESTED** [review]

File: pallets/democracy/src/lib.rs [hunk 6]
```diff
@@ -179,6 +179,7 @@ pub mod weights;
 use crate::traits::DemocracyHooks;
 pub use conviction::Conviction;
 pub use pallet::*;
+pub use sp_std::vec;
```
Comment (enthusiastmartin) [inline pallets/democracy/src/lib.rs:182]: wait ..what? why ?

Comment (mrq1911) [reply pallets/democracy/src/lib.rs:182]: my question exactly

File: pallets/stableswap/src/trade_execution.rs [hunk 7]
```diff
@@ -30,10 +71,64 @@ impl<T: Config> TradeExecution<T::RuntimeOrigin, T::AccountId, T::AssetId, Balan
 	) -> Result<Balance, ExecutorError<Self::Error>> {
 		match pool_type {
 			PoolType::Stableswap(pool_id) => {
-				let (amount_in, _) = Self::calculate_in_amount(pool_id, asset_in, asset_out, amount_out)
-					.map_err(ExecutorError::Error)?;
+				if asset_out == pool_id {
+					//I wanna buy 500 shares, how much luqidity i need provide to get 500 shares
```
Comment (mrq1911) [inline pallets/stableswap/src/trade_execution.rs:75]: remove comment

File: pallets/stableswap/src/trade_execution.rs [hunk 8]
```diff
@@ -65,7 +178,17 @@ impl<T: Config> TradeExecution<T::RuntimeOrigin, T::AccountId, T::AssetId, Balan
 	) -> Result<(), ExecutorError<Self::Error>> {
 		match pool_type {
 			PoolType::Stableswap(pool_id) => {
-				Self::buy(who, pool_id, asset_out, asset_in, amount_out, max_limit).map_err(ExecutorError::Error)
+				if asset_out == pool_id {
+					Self::add_liquidity_shares(who, pool_id, amount_out, asset_in, max_limit)
+						.map_err(ExecutorError::Error)
+				} else if asset_in == pool_id {
+					let shares_amount = max_limit;
```
Comment (mrq1911) [inline pallets/stableswap/src/trade_execution.rs:185]: inline

File: pallets/stableswap/src/trade_execution.rs [hunk 9]
```diff
@@ -49,7 +144,25 @@ impl<T: Config> TradeExecution<T::RuntimeOrigin, T::AccountId, T::AssetId, Balan
 	) -> Result<(), ExecutorError<Self::Error>> {
 		match pool_type {
 			PoolType::Stableswap(pool_id) => {
-				Self::sell(who, pool_id, asset_in, asset_out, amount_in, min_limit).map_err(ExecutorError::Error)
+				if asset_in == pool_id {
+					Self::remove_liquidity_one_asset(who, pool_id, asset_out, amount_in, min_limit)
+						.map_err(ExecutorError::Error)
+				} else if asset_out == pool_id {
+					let decimals = T::AssetInspection::decimals(asset_in)
```
Comment (mrq1911) [inline pallets/stableswap/src/trade_execution.rs:151]: do you really have to specify decimals in extrinsic?

Comment (enthusiastmartin) [reply pallets/stableswap/src/trade_execution.rs]: removed

File: runtime/hydradx/src/assets.rs (hunk 4 above)
Comment (enthusiastmartin) [reply runtime/hydradx/src/assets.rs]: added.

**Review by mrq1911 — APPROVED** [review]

**Pull request merged**